ovs/lib/helpers/decorators.py

```diff
@@ -122,380 +172,885 @@ def _ensure_single(task_name, mode, extra_task_names=None, global_timeout=300, c
+
+    def callback_task(self, message=''):
+        # type: (str) -> None
+        """
+        Discard the execution of a task by raising EnsureSingleDoCallBack
+        :return: None
+        :rtype: NoneType
+        :raises: EnsureSingleDoCallBack
+        """
+        self.unittest_set_state_callback()
+        self.run_hook('before_callback')
+        callback_message = message or 'Execution of task {0} in progress, execute callback function instead'.format(self.ensure_single_container.task_name)
+        self.logger.info(self.message.format(callback_message))
+        raise EnsureSingleDoCallBack()
+
+    def do_discard(self):
+        # type: () -> None
+        """
+        Discard the task
+        The easiest way to discard is to never allow the celery task to run by returning prematurely
+        :return: None
+        """
+        self.run_hook('after_discard')
+        return None
+
+    def do_callback(self, *args, **kwargs):
+        # type: (*any, **any) -> any
+        """
+        Execute the given callback (if any)
+        """
+        if not self.ensure_single_container.callback:
+            error_message = 'No callback passed and callback invocation is requested!'
+            self.logger.error(error_message)
+            raise ValueError(error_message)
+
+        # This log message might be off if when running it outside the decorators, which isn't supposed to be done
+        callback_message = 'Execution of task {0} in progress, executing callback function'.format(self.ensure_single_container.task_name)
+        self.logger.info(self.message.format(callback_message))
+        result = self.ensure_single_container.callback(*args, **kwargs)
+        self.run_hook('after_callback')
+        return result
+
+    def run_hook(self, hook):
+        # type: (str) -> None
+        """
+        Run a hook
+        :param hook: Hook identifier to run
+        :type hook: str
+        :return: None
+        :rtype: NoneType
+        """
+        if self.unittest_mode:
+            hook_func = self.ensure_single_container.hooks.get(hook)
+            if not hook_func:
+                return
+            hook_func()
+
+    def gather_run_time_info(self):
+        # type: () -> None
+        """
+        Retrieve some information only acquirable when going to execute the code
+        """
+        self.now = '{0}_{1}'.format(int(time.time()), ''.join(random.choice(string.ascii_letters + string.digits) for _ in range(10)))
+        self.thread_name = threading.current_thread().getName()
+        self.unittest_mode = os.environ.get('RUNNING_UNITTESTS') == 'True'
+        if self.unittest_mode:
+            self.message = 'Ensure single {0} mode - ID {1} - {2} - {{0}}'.format(self.ensure_single_container.mode, self.now, self.thread_name)
+        else:
+            self.message = 'Ensure single {0} mode - ID {1} - {{0}}'.format(self.ensure_single_container.mode, self.now)
+
+    def unittest_set_state(self, state, value):
```

Comment:
Perhaps set None as default for `value`, or does this have flaws?

Reply:
No need